MIP22: fixes

xmxanuel opened the pull request:
**Content**
- fix events to custom events
- rename ONE to RAY
- changed tranche approval `uint(-1)`
- rm pool approval
- rename owner to operator
- adjusted unit tests

-  filing functions for vow, end and join
-  return funds to operator after end has settled
-  add revert messages
-  operator can no longer call tell

xmxanuel committed:
fix: approval for tranche

xmxanuel committed:
decativated rpc tests

xmxanuel committed:
Merge pull request #20 from manuelpolzhofer/fixApprovals

fix: approval for tranche contract

xmxanuel committed:
fix: custom events

xmxanuel committed:
fix: rename owner to operator

xmxanuel committed:
Merge pull request #21 from manuelpolzhofer/fixEvents

fix: MIP22 events

ilinzweilin committed:
run unit tests instead of rpc tests in github workflow

ilinzweilin committed:
fix: remove comment

xmxanuel committed:
Merge pull request #22 from centrifuge/fix/tests

run unit tests instead of rpc tests in github workflow

MrChico commented on src/mgr.sol:
```diff
@@ -127,20 +150,20 @@ contract TinlakeManager is LibNote {
         ilk = ilk_;
         wards[msg.sender] = 1;
 
```
```suggestion
emit Rely(msg.sender); 
```

xmxanuel replied:
I adopted it from the MIP21
https://github.com/livnev/rwa-example/pull/8/files#diff-20efe103664fbcfe47bafc2b8deb1cc4fcb9939f45fe32d91c2b0dee0ddc551aR31

MrChico replied:
see https://github.com/centrifuge/tinlake-maker-lib/pull/18#discussion_r565448088

ilinzweilin committed:
resolve conflicts with base branch

ilinzweilin committed:
fix: add comments

ilinzweilin committed:
Merge branch 'MIP22_fixes' into fix/merge-conflicts

ilinzweilin committed:
Merge pull request #24 from centrifuge/fix/merge-conflicts

fix: merge conflicts

MrChico committed:
- filing functions for vow, end and join
- return funds to operator after end has settled
- add revert messages
- a few extra tests

MrChico committed:
Apply suggestions from code review

MrChico committed:
add end mock

ilinzweilin committed:
Merge pull request #25 from centrifuge/cagefiles

fixes

gbalabasquer commented on src/mgr.sol:
```diff
@@ -70,38 +69,62 @@ interface RedeemLike {
-    mapping (address => uint256) public wards;
-    function rely(address usr) external note auth { require(live, "TinlakeMgr/not-live"); wards[usr] = 1; }
-    function deny(address usr) external note auth { require(live, "TinlakeMgr/not-live"); wards[usr] = 0; }
+     mapping (address => uint256) public wards;
```
nit: extra spacing

gbalabasquer commented on src/mgr.sol:
```diff
@@ -70,38 +69,62 @@ interface RedeemLike {
-    // The owner manages the cdp, but is not authorized to call kick or cage.
-    address public owner;
+    // Events
+    event Rely(address usr);
```
I think we could set all addresses as `indexed` in the different events.

gbalabasquer replied:
I've seen `indexed` has been added for several events but still missing for `Rely` and `Deny`

gbalabasquer commented on src/mgr.sol:
```diff
@@ -70,38 +69,62 @@ interface RedeemLike {
+    event Sink(uint ink, uint tab);
+    event Recover(uint recovered, uint payBack);
+    event Cage();
+    event File(bytes32 what, address data);
```
`what` might be `indexed` as well

gbalabasquer commented on src/mgr.sol:
```diff
@@ -70,38 +69,62 @@ interface RedeemLike {
+    // Events
+    event Rely(address usr);
+    event Deny(address usr);
+    event Draw(uint wad);
```
nit: if possible make all the `uint` to `uint256`.

gbalabasquer replied:
Same for `int256`. I think some of them that we had changed, have been overwritten in this PR.

gbalabasquer commented on src/mgr.sol:
```diff
@@ -159,64 +186,80 @@ contract TinlakeManager is LibNote {
-        owner = newOwner;
+    function setOperator(address newOperator) external operatorOnly  {
+        operator = newOperator;
+        emit NewOperator(newOperator);
```
Hmm maybe better calling it as `SetOperator`

gbalabasquer commented on src/mgr.sol:
```diff
@@ -159,64 +186,80 @@ contract TinlakeManager is LibNote {
+        require(live, "TinlakeManager/not-live");
+        emit File(what, data);
+        if (what == "vow") vow = data;
+        else if (what == "join") daiJoin = JoinLike(data);
```
I'd suggest call the param also `daiJoin`.

gbalabasquer commented on src/mgr.sol:
```diff
@@ -110,38 +133,42 @@ contract TinlakeManager is LibNote {
-        gem.approve(pool_, uint256(-1));
-        gem.approve(tranche, uint256(-1));
+
+        tranche = tranche_;
```
I know this comes from our comments, so actually directing this comment to @godsflaw and @brianmcmichael. Is it really worth it to save the variable to do partial approvals on `tell`? I mean if we can actually verify that `tranche` can only take `gem` when `tell` is called by governance we could avoid having to store this one.

xmxanuel replied:
The tranche `redeemOrder` can only be triggered by Maker governance.  This is the normal way for DROP token holder to redeem their tokens. 
The mgr calls the `operator.redeemOrder`
redeemOrder in operator
https://github.com/centrifuge/tinlake/blob/develop/src/lender/operator.sol#L67
redeemOrder in tranche
https://github.com/centrifuge/tinlake/blob/develop/src/lender/tranche.sol#L128

I think we should just keep the `amount` approval then there is definitely no way that Tinlake
can access it.

In Tinlake we use a contract called `clerk` (operator in mgr terminology)  to interact with the `mgr`.
For all methods we require the following to be true:
```javascript
 function mkrActive() public view returns (bool) {
        return mgr.safe() && mgr.glad() && mgr.live();
    }
```
See: https://github.com/centrifuge/tinlake/blob/develop/src/lender/adapters/mkr/clerk.sol#L117

gbalabasquer commented on src/mgr.sol:
```diff
@@ -159,64 +186,80 @@ contract TinlakeManager is LibNote {
-    function tell() public note {
-        require(safe);
-        require(wards[msg.sender] == 1 || (msg.sender == owner && !live), "TinlakeManager/not-authorized");
+    function tell() public auth {
```
It seems `tell` can't be called anymore by the operator if shutdown happens. Is that an intentional change?
If this change was actually wrong, couldn't we even make that this is called by anyone (if shutdown happens)?

xmxanuel replied:
Yes, previously the only way to set the live flag to false was through the `migrate` call and by calling `cage` but cage requires a tell call beforehand. 

In the end the operator doesn't to have the power to trigger the `tell`.

That was the reason why @MrChico removed it.

@MrChico , correct?

gbalabasquer replied:
Hmm not really, `cage` can be called by anyone if shutdown happens in the MCD system. Prior to this PR, the operator could run `tell` if the `!live`.

MrChico replied:
Yes, removing the ability for the operator to call `tell` was intentional. The operator can de-risk the position in other ways, and removing this privilege for the operator reduces some complexity.

Note that `mgr.cage()` requires `!glad` which implies that `tell` has already been called.

gbalabasquer replied:
Ah yes, you are right, so `cage` can't be called if governance hasn't previously called `tell`.
Tbh I need to draw a states flow chart to understand well all these cases.

MrChico replied:
Yeah, we should probably add one to the README

xmxanuel committed:
fix: uint to uint256 and indexed events

xmxanuel committed:
added state diagram to readme

xmxanuel committed:
dapp uninstall tinlake

xmxanuel committed:
dapp install https://github.com/centrifuge/tinlake

gbalabasquer commented on src/mgr.sol:
```diff
@@ -95,18 +95,18 @@ contract TinlakeManager {
     // Events
     event Rely(address usr);
     event Deny(address usr);
```
Conversation got resolved but the `indexed` is still missing for `Rely` and `Deny`

xmxanuel replied:
fixed

xmxanuel committed:
add event index for rely and deny

hexonaut commented on src/mgr.sol:
```diff
@@ -110,33 +133,37 @@ contract TinlakeManager is LibNote {
         pool = RedeemLike(pool_);
 
         ilk = ilk_;
         wards[msg.sender] = 1;
```
Needs `emit Rely(msg.sender);` after this.

godsflaw replied:
```suggestion
        wards[msg.sender] = 1;
        emit Rely(msg.sender);
```

hexonaut commented on src/mgr.sol:
```diff
@@ -110,33 +133,37 @@ contract TinlakeManager is LibNote {
+        daiJoin = JoinLike(daiJoin_);
+        vow = vow_;
 
         gem = GemLike(drop_);
```
Should add `require(gem.decimals() == dec)` for defensive programming.

godsflaw replied:
```suggestion
        gem = GemLike(drop_);
        require(gem.decimals() == dec, "TinlakeMgr/decimals-dont-match");
```

hexonaut reviewed (changes requested):
Just some minor adjustments.

hexonaut commented:
In general I would like to see some additional tests against the actual dss code instead of just mocks. Specifically the liquidation and write-off processes.

xmxanuel commented:
@hexonaut thanks for your review you can find system tests in this repo:
https://github.com/centrifuge/tinlake-maker-tests

We have additional tests in the Tinlake core repo (https://github.com/centrifuge/tinlake/tree/develop/src/test/system/lender/mkr) where Maker is mocked but we re-run the same tests in `tinlake-maker-tests` with actual Maker contracts and the `mgr`

You can also find additional liquidation and write-offs tests in the test repo

godsflaw commented on src/mgr.sol:
```diff
@@ -17,7 +17,6 @@
 // along with this program.  If not, see <https://www.gnu.org/licenses/>.
 
 pragma solidity >=0.5.12;
```
```suggestion
pragma solidity 0.5.12;
```

godsflaw reviewed (approved):
LGTM on accepted changes

xmxanuel committed:
Update src/mgr.sol

Co-authored-by: Christopher Mooney <chris@dod.net>

xmxanuel committed:
Update src/mgr.sol

Co-authored-by: Christopher Mooney <chris@dod.net>

xmxanuel committed:
Update src/mgr.sol

Co-authored-by: Christopher Mooney <chris@dod.net>

lucasvo commented:
Closing this PR as we're changing strategy a bit.
